instrumentation/opentelemetry-instrumentation-click/src/opentelemetry/instrumentation/click/__init__.py

```diff
@@ -0,0 +1,105 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Instrument `click`_ CLI applications.
+
+.. _click: https://pypi.org/project/click/
+
+Usage
+-----
+
+.. code-block:: python
+
+    import click
+    from opentelemetry.instrumentation.click import ClickInstrumentor
+
+    ClickInstrumentor().instrument()
+
+    @click.command()
+    def hello():
+       click.echo(f'Hello world!')
+
+    if __name__ == "__main__":
+        hello()
+
+API
+---
+"""
+
+from functools import partial
+from logging import getLogger
+from typing import Collection
+
+import click
+from wrapt import wrap_function_wrapper
+
+from opentelemetry import trace
+from opentelemetry.instrumentation.click.package import _instruments
+from opentelemetry.instrumentation.click.version import __version__
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.utils import (
+    unwrap,
+)
+from opentelemetry.semconv.attributes.error_attributes import ERROR_TYPE
+from opentelemetry.trace.status import StatusCode
+
+_logger = getLogger(__name__)
+
+
+def _command_invoke_wrapper(wrapped, instance, args, kwargs, tracer):
+    # Subclasses of Command include groups and CLI runners, but
+    # we only want to instrument the actual commands which are
+    # instances of Command itself.
+    if instance.__class__ != click.Command:
+        return wrapped(*args, **kwargs)
+
+    ctx = args[0]
+    span_name = ctx.info_name
+    span_attributes = {}
+
+    with tracer.start_as_current_span(
+        name=span_name,
+        kind=trace.SpanKind.INTERNAL,
+        attributes=span_attributes,
+    ) as span:
+        try:
+            return wrapped(*args, **kwargs)
+        except Exception as exc:
+            span.set_status(StatusCode.ERROR, str(exc))
+            span.set_attribute(ERROR_TYPE, exc.__class__.__qualname__)
```

Comment:
span.is_recording() check would be good here.

Reply:
There may be a chance that adding `span.is_recording()` conditionals for single attributes causes more overhead than just adding the attribute itself :sweat_smile: